libraries/AP_NavEKF3/AP_NavEKF3_MagFusion.cpp

```diff
@@ -855,70 +843,82 @@ void NavEKF3_core::FuseMagnetometer()
- * It is not as robust to magnetometer failures.
- * It is not suitable for operation where the horizontal magnetic field strength is weak (within 30 degrees latitude of the magnetic poles)
- *
- * The following booleans are passed to the function to control the fusion process:
- *
- * usePredictedYaw -  Set this to true if no valid yaw measurement will be available for an extended periods.
- *                    This uses an innovation set to zero which prevents uncontrolled quaternion covariance
- *                    growth or if available, a yaw estimate from the Gaussian Sum Filter.
- * UseExternalYawSensor - Set this to true if yaw data from an external yaw sensor (GPS or external nav) is being used instead of the magnetometer.
+ * Returns true if the fusion was successful
 */
-void NavEKF3_core::fuseEulerYaw(bool usePredictedYaw, bool useExternalYawSensor)
+bool NavEKF3_core::fuseEulerYaw(yawFusionMethod method)
 {
     const float &q0 = stateStruct.quat[0];
     const float &q1 = stateStruct.quat[1];
     const float &q2 = stateStruct.quat[2];
     const float &q3 = stateStruct.quat[3];
 
-    // external yaw available check
-    bool canUseGsfYaw = false;
-    float gsfYaw, gsfYawVariance, velInnovLength;
-    if (usePredictedYaw && yawEstimator != nullptr) {
-        canUseGsfYaw = yawEstimator->getYawData(gsfYaw, gsfYawVariance) &&
-                       is_positive(gsfYawVariance) &&
-                       gsfYawVariance < sq(radians(GSF_YAW_ACCURACY_THRESHOLD_DEG)) &&
-                       (assume_zero_sideslip() || (yawEstimator->getVelInnovLength(velInnovLength) && velInnovLength < frontend->maxYawEstVelInnov));
+    float gsfYaw, gsfYawVariance;
+    if (method == yawFusionMethod::GSF) {
+        bool canUseGsfYaw;
+        if (yawEstimator != nullptr) {
+            float velInnovLength;
+            canUseGsfYaw = yawEstimator->getYawData(gsfYaw, gsfYawVariance) &&
+                        is_positive(gsfYawVariance) &&
+                        gsfYawVariance < sq(radians(GSF_YAW_ACCURACY_THRESHOLD_DEG)) &&
+                        (assume_zero_sideslip() || (yawEstimator->getVelInnovLength(velInnovLength) && velInnovLength < frontend->maxYawEstVelInnov));
+        } else {
+            canUseGsfYaw = false;
+        }
+        if (!canUseGsfYaw) {
+            return false;
+        }
     }
 
     // yaw measurement error variance (rad^2)
     float R_YAW;
-    if (canUseGsfYaw) {
+    switch (method) {
+    case yawFusionMethod::EXTERNAL:
+        R_YAW = sq(yawAngDataDelayed.yawAngErr);
+        break;
+
+    case yawFusionMethod::GSF:
         R_YAW = gsfYawVariance;
-    } else if (!useExternalYawSensor) {
+        break;
+
+    case yawFusionMethod::STATIC:
+        R_YAW = sq(yawAngDataStatic.yawAngErr);
+        break;
+
+    case yawFusionMethod::MAGNETOMER:
+    case yawFusionMethod::PREDICTED:
+    default:
```

Comment:
`default:` probably isn't appropriate here - we want the compiler to throw an error in the case a new `yawFusionMethod` isn't handled, and I don't think `method` can come from anywhere but the relevant enumeration.

Reply:
If I remove that default  then i get a compiler warning saying that R_YAW could be used uninitialised.

Reply:
how very odd.  I've confirmed that I also see the compiler warning but I don't understand why so I'll leave it to PeterB to answer.  By the way I noticed there are some missing breaks below the "default:" but it doesn't make any functional difference and doesn't resolve the compiler warning.  I still think we should always have a break though even for the last case.

Reply:
There's a commit at the end of this branch that adds the missing break statements in case that's useful: https://github.com/rmackay9/rmackay9-ardupilot/commits/prise-ekf3-yaw-spin-fix